core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,263 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.SerializableMap;
+import org.apache.iceberg.util.SerializableSupplier;
+
+/**
+ * SnapshotOperations abstracts access to snapshots for table metadata. This allows a subset of
+ * snapshots/refs to be loaded initially for operations that do not require all to be present (e.g.
+ * table scans of the current snapshot).
+ *
+ * <p>In the event that snapshots/refs need to be accessed, the provided supplier will be invoked,
+ * which must provide all addressable snapshots for the table.
+ */
+class SnapshotOperations implements Serializable {
+  private List<Snapshot> snapshots;
+  private final SerializableSupplier<List<Snapshot>> snapshotsSupplier;
+  private Map<Long, Snapshot> snapshotsById;
+  private Map<String, SnapshotRef> refs;
+  private final SerializableSupplier<Map<String, SnapshotRef>> refsSupplier;
+  private boolean loaded;
+
+  private SnapshotOperations(
+      List<Snapshot> snapshots,
+      SerializableSupplier<List<Snapshot>> snapshotsSupplier,
+      Map<String, SnapshotRef> refs,
+      SerializableSupplier<Map<String, SnapshotRef>> refsSupplier) {
+    this.snapshots = ImmutableList.copyOf(snapshots);
+    this.snapshotsSupplier = snapshotsSupplier;
+    this.refs = SerializableMap.copyOf(refs);
+    this.refsSupplier = refsSupplier;
+
+    this.snapshotsById = indexSnapshotsById(snapshots);
+  }
+
+  List<Snapshot> snapshots() {
+    ensureLoaded();
+
+    return snapshots;
+  }
+
+  Snapshot snapshot(long id) {
+    if (!snapshotsById.containsKey(id)) {
+      ensureLoaded();
+    }
+
+    return snapshotsById.get(id);
+  }
+
+  boolean contains(long id) {
+    if (!snapshotsById.containsKey(id)) {
+      ensureLoaded();
+    }
+
+    return snapshotsById.containsKey(id);
+  }
+
+  Map<String, SnapshotRef> refs() {
+    return refs;
+  }
+
+  SnapshotRef ref(String name) {
+    return refs.get(name);
+  }
+
+  private void ensureLoaded() {
```

Comment:
do we care about parallel access here? just wondering whether this should be `synchronized` here